### PR TITLE
API / UI modes from YML config

### PR DESCRIPTION
**Pull request by newstler**
Seems I cannot push to your repository, so I had to push to mine :)

Here I create `modes.yml`, which now contains gems to remove, but I should also add files to remove and update file removing code in `bin/configuration`. Also should add code to delete that `modes.yml` after config.

Overall it feels pretty smelly code, to be honest. I've been learning how is it done in Rails `new`, of course there are a lot of magic, but I feel like as a framework we should adapt their approach. Of course, that could be done later if there is a need in that.

Also I noticed weird code that just deletes `.github/workflows/main.yml`, why are we shipping that file then? (Probably should be in another task)

I also rearranged code into callable methods so it would be better for reading. In the end I think it all should be placed in some class or module to keep configuration clean and understandable.

### CODE REVIEW

**Comment by gazayas** [comment]
Hey @newstler, LGTM!

> Also I noticed weird code that just deletes .github/workflows/main.yml, why are we shipping that file then? (Probably should be in another task)

We had the `edge` branch set up before which the starter repository used, but new Bullet Applications don't need them so [the worfklow was deleted here](https://github.com/bullet-train-co/bullet_train/commit/b46ce052c1dddb5dd1ec19a0f11149a3d233f7f9). I wrote up an issue [here too](https://github.com/bullet-train-co/bullet_train/issues/714).

> Overall it feels pretty smelly code, to be honest. I've been learning how is it done in Rails new, of course there are a lot of magic, but I feel like as a framework we should adapt their approach. Of course, that could be done later if there is a need in that.

Yeah, this is why I initially wanted to do something like this:
```
> git clone --branch api-only git@github.com:bullet-train-co/bullet_train.git application_name
```

It would be nice to make things as uniform as possible with `rails new`, but I think it will be pretty difficult in our use case considering how Bullet Train is set up.

Thanks for this!

File: bin/configure [hunk 1]
```diff
@@ -1,115 +1,82 @@
...
-    end
-  else
-    target_file_content = File.open(file).read
-    target_file_content.gsub!(before, after)
-  end
-  File.open(file, "w+") do |f|
-    f.write(target_file_content)
-  end
-end
+def install_necessary_gems!
```
Comment (gazayas) [inline bin/configure]: Any reason we use `!` for these methods?

Comment (newstler) [reply bin/configure]: Just to differ "step" methods from the secondary ones. Not sure if that's right but it helps in reading code for me personally. Can switch them to normal if we follow the Ruby convention to denote that a method can be "dangerous".

**Comment by gazayas** [comment]
I didn't mention this earlier, but I can go and merge this into my branch, just hadn't done so yet because it's still a draft.

**Comment by newstler** [comment]
Yep I'm working on files in YML, planned to do it earlier but was disturbed by weekend 😅 Should finish today or tomorrow

**Comment by newstler** [comment]
- Done with files and gems (please check if I forgot anything).
- Tried to update `edit_ci!` method, but it appears I am not very familiar with how this CI config works. I noticed there lines like `browser-tools: circleci/browser-tools@1.4.1`, `restore_yarn_cache` and so on, but it feels like it's been used everywhere, even
```
  'Database Schema Check':
    docker:
      - <<: *ruby_node_browsers_docker_image
```
which is pretty weird to me, but maybe I don't know something.
- I realised that it's impossible to create `Team` or `User` from API, so probably all the API-mode app would be unusable 😅 Should we add those actions to `bullet_train-api` gem, how do you think?

File: bin/configure (hunk 1 above)
Comment (newstler) [reply bin/configure]: Leaved bang only for dangerous methods.

**Comment by gazayas** [comment]
## Creating Users and Teams
> I realised that it's impossible to create Team or User from API, so probably all the API-mode app would be unusable sweat_smile Should we add those actions to bullet_train-api gem, how do you think?

Yes, this is something I was thinking about. Check out the conversation [here on Discord](https://discord.com/channels/836637622432170028/1100307029282787449/1105443487203209287) about authenticating users with Doorkeeper via their email and password. I think we'll have to take a similar approach when it comes to creating Users and Teams, so it's definitely something we'll need to sift through.

Since this would be a pretty big design decision, I think we'll have to wait and see what Andrew would like to do about this. cc/ @andrewculver I'm tagging you since this is my forked repository and I'm not sure if you get notifications from here :pray:.

In the meantime I think the best we can do is set up the repository to make it api-ready, and then think through ways of how to create a User/Team via the API.

## Front End dependencies in CI
I don't think we really considered having an API-only setup when creating these CI tests, so I think this is something we can think about later. We definitely don't need to restore the yarn cache (or run `yarn install` for that matter) or use browser tools if that's the case though.

## Review
Besides that, just give me some time and I'll try to do a more thorough review! Thanks for all of this!

**Comment by gazayas** [comment]
@newstler This looks great! Compartmentalizing the logic into different methods makes things a lot easier to read, and it reminds me of the setup we have in the [Action Models Transformer](https://github.com/bullet-train-pro/bullet_train-action_models/blob/4b896a0099df1fa6b4eaec9c69ed200986f08719/lib/scaffolding/action_model_transformer.rb#L240-L249).

After running the setup, I have to run `bundle update capybara` (not sure if it's just me), and I'm getting a couple of errors related to `config.assets` in the `application.rb` and `development.rb` config files. Any idea as to what could be causing this?

<details>
  <summary>NoMethodError: undefined method `assets'</summary>

```
NoMethodError: undefined method `assets' for #<Rails::Application::Configuration:0x00007feba8f78c08 @root=#<Pathname:/home/gazayas/work/bt/bullet_train>, @generators=#<Rails::Configuration::Generators:0x00007feba9663220 @aliases={}, @options={:rails=>{:orm=>:active_record, :test_framework=>:test_unit, :integration_tool=>:test_unit, :system_tests=>:test_unit}, :active_record=>{:migration=>true, :timestamps=>true}, :test_unit=>{:fixture=>true, :fixture_replacement=>nil}}, @fallbacks={}, @templates=[], @colorize_logging=true, @api_only=false, @hidden_namespaces=[], @after_generate_callbacks=[]>, @middleware=#<Rails::Configuration::MiddlewareStackProxy:0x00007feba78fffa8 @operations=[], @delete_operations=[]>, @javascript_path="javascript", @encoding=#<Encoding:UTF-8>, @allow_concurrency=nil, @consider_all_requests_local=false, @filter_parameters=[], @filter_redirect=[], @helpers_paths=[], @hosts=[".localhost", #<IPAddr: IPv4:0.0.0.0/0.0.0.0>, #<IPAddr: IPv6:0000:0000:0000:0000:0000:0000:0000:0000/0000:0000:0000:0000:0000:0000:0000:0000>], @host_authorization={}, @public_file_server=#<ActiveSupport::OrderedOptions {:enabled=>true, :index_name=>"index"}>, @force_ssl=false, @ssl_options={:hsts=>{:subdomains=>true}}, @session_store=nil, @time_zone="UTC", @beginning_of_week=:monday, @log_level=:debug, @cache_store=[:file_store, "/home/gazayas/work/bt/bullet_train/tmp/cache/"], @railties_order=[:all], @relative_url_root=nil, @reload_classes_only_on_change=true, @file_watcher=ActiveSupport::FileUpdateChecker, @exceptions_app=nil, @autoflush_log=true, @log_formatter=#<ActiveSupport::Logger::SimpleFormatter:0x00007feba78f66b0 @datetime_format=nil>, @eager_load=nil, @secret_key_base=nil, @api_only=false, @debug_exception_response_format=nil, @x=#<Rails::Application::Configuration::Custom:0x00007feba78f65e8 @configurations={}>, @enable_dependency_loading=false, @read_encrypted_secrets=false, @content_security_policy=nil, @content_security_policy_report_only=false, @content_security_policy_nonce_generator=nil, @content_security_policy_nonce_directives=nil, @require_master_key=false, @loaded_config_version=7.0, @credentials=#<ActiveSupport::OrderedOptions {:content_path=>#<Pathname:/home/gazayas/work/bt/bullet_train/config/credentials.yml.enc>, :key_path=>#<Pathname:/home/gazayas/work/bt/bullet_train/config/master.key>}>, @disable_sandbox=false, @add_autoload_paths_to_load_path=true, @permissions_policy=nil, @rake_eager_load=false, @server_timing=false>
/home/gazayas/work/bt/bullet_train/config/application.rb:30:in `<class:Application>'
/home/gazayas/work/bt/bullet_train/config/application.rb:13:in `<module:ProjectGabe>'
/home/gazayas/work/bt/bullet_train/config/application.rb:12:in `<main>'
/home/gazayas/work/bt/bullet_train/Rakefile:4:in `require_relative'
/home/gazayas/work/bt/bullet_train/Rakefile:4:in `<main>'
/home/gazayas/.rbenv/versions/3.2.2/bin/bundle:25:in `load'
/home/gazayas/.rbenv/versions/3.2.2/bin/bundle:25:in `<main>'
```

</details>

**Comment by newstler** [comment]
> This looks great! Compartmentalizing the logic into different methods makes things a lot easier to read, and it reminds me of the setup we have in the [Action Models Transformer](https://github.com/bullet-train-pro/bullet_train-action_models/blob/4b896a0099df1fa6b4eaec9c69ed200986f08719/lib/scaffolding/action_model_transformer.rb#L240-L249).

Thank you, never seen Action Models Transformer code, but that is actually what I do do in any script — just make step methods and then call them so it becomes more readable. It also allows simplify and improve readability of conditional statements.

> After running the setup, I have to run bundle update capybara (not sure if it's just me), and I'm getting a couple of errors related to config.assets in the application.rb and development.rb config files. Any idea as to what could be causing this?

Yes, I think I know what to do, will try to fix today.

**Comment by newstler** [comment]
Replaced the code for the API mode in 
- `config/application.rb`
- `config/environments/development.rb`
- `config/environments/test.rb`
- `config/environments/production.rb`

Also refactored `config/environments/production.rb` a little to remove duplications.

**Pull request merged**